AK/URLParser.cpp

```diff
@@ -39,17 +39,44 @@ static void report_validation_error(SourceLocation const& location = SourceLocat
     dbgln_if(URL_PARSER_DEBUG, "URLParser::basic_parse: Validation error! {}", location);
 }
 
+// https://url.spec.whatwg.org/#concept-opaque-host-parser
 static Optional<URL::Host> parse_opaque_host(StringView input)
 {
+    // 1. If input contains a forbidden host code point, host-invalid-code-point validation error, return failure.
     auto forbidden_host_characters_excluding_percent = "\0\t\n\r #/:<>?@[\\]^|"sv;
     for (auto character : forbidden_host_characters_excluding_percent) {
         if (input.contains(character)) {
             report_validation_error();
             return {};
         }
     }
-    // FIXME: If input contains a code point that is not a URL code point and not U+0025 (%), validation error.
-    // FIXME: If input contains a U+0025 (%) and the two code points following it are not ASCII hex digits, validation error.
+
+    // 2. If input contains a code point that is not a URL code point and not U+0025 (%), invalid-URL-unit validation error.
+    for (auto character : input) {
+        if (!is_url_code_point(character) && character != '%') {
+            report_validation_error();
+            return {};
```

Comment:
Just realized my implementation of spec steps 2 and 3 are wrong as the[ spec explains](https://url.spec.whatwg.org/#validation-error) that a validation error does not mean a failure. Fixing this hurts testability, however, as I can no longer use `URL::is_valid()` to check if a validation error has occurred. `report_validation_error()` just prints a debug log of where the error occurred (which seems fine as the spec just wants us to "report them somewhere").

I currently see 2 solutions. We could possibly define the `URL_PARSER_DEBUG` macro in the test code and check if the debug log is printed. This will only work if our testing framework would allow us to capture stdout/stderr (not sure where `dbgln_if()` writes to). Another solution would be to change the signature and body of `report_validation_error()` such that we can somehow detect in the testing code that a validation error has occurred.

Reply:
Okay, had a chance to look into this properly

It's not worthwhile adding tests for non-functional code. When I say we should ideally be adding tests, I mean for when we're making fixes / implementing this (where it is possible to do so). Reporting validation failure is really only useful for developer debug ATM for someone working on URL parsing, and I think it may very well may just stay that way forever.

Speaking of which - I'm thinking that it may not even worthwhile implementing those two steps. All of the other places that validation errors are reported are super simple, cheap checks. Those two checks aren't sotrivial. I'm thinking it may just be fine to leave these steps two steps unimplemented and just a comment about that.

By the way, these tests can live in TestURL.cpp - that's pretty much where all of the other similar tests are :^)

Reply:
Thanks for looking into this and for your insight! I agree that these 2 steps seem to be more useful for debugging efforts rather than providing value to the end user. Both steps also run in linear time so I see your point of them not being cheap.

I'll keep the first commit since `is_url_code_point()` is used frequently in `URLParser`, but for my second commit I'll remove my spec implementation code and just add a note beneath the 2 spec step comments👍

Reply:
Made the necessary changes, when you get the time let me know if it needs improvements!